huaweicloud/services/servicestage/resource_huaweicloud_servicestage_application.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/chnsz/golangsdk/openstack/servicestage/v2/components"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+)
+
+func ResourceApplication() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceApplicationCreate,
+		ReadContext:   resourceApplicationRead,
+		UpdateContext: resourceApplicationUpdate,
+		DeleteContext: resourceApplicationDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"region": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[A-Za-z]([\w-]*[A-Za-z0-9])?$`),
+						"The name must start with a letter and end with a letter or digit, and can only contain "+
+							"letters, digits, underscores (_) and hyphens (-)."),
+					validation.StringLenBetween(2, 64),
+				),
+			},
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringLenBetween(0, 128),
+			},
+			"enterprise_project_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"environment": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"variable": {
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": {
+										Type:     schema.TypeString,
+										Required: true,
+										ValidateFunc: validation.All(
+											validation.StringMatch(regexp.MustCompile(`^[A-Za-z-_][\w-.]*$`),
+												"The name can only contain letters, digits, underscores (_), "+
+													"hyphens (-) and dots (.), and cannot start with a digit or dot."),
+											validation.StringLenBetween(1, 64),
+										),
+									},
+									"value": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validation.StringLenBetween(1, 2048),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"component_ids": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func createEnvironmentVariables(client *golangsdk.ServiceClient, appId string, envSet *schema.Set) error {
+	for _, env := range envSet.List() {
+		envMap := env.(map[string]interface{})
+		variables := make([]applications.Variable, envMap["variable"].(*schema.Set).Len())
+		for i, v := range envMap["variable"].(*schema.Set).List() {
```

Comment:
suggest to use a variable to save `envMap["variable"].(*schema.Set)`

Reply:
done.